bigquery_schema_generator/generate_schema.py

```diff
@@ -340,31 +369,75 @@ def merge_schema_entry(
                 )
             return old_schema_entry
 
+        new_mode = self.merge_mode(old_schema_entry,
+                                   new_schema_entry,
+                                   base_path)
+        if new_mode is None:
+            return None
+        new_schema_entry['info']['mode'] = new_mode
+
+        # For all other types...
+        if old_type != new_type:
+            # Check that the converted types are compatible.
+            candidate_type = convert_type(old_type, new_type)
+            if not candidate_type:
+                self.log_error(
+                    f'Ignoring field with mismatched type: '
+                    f'old=({old_status},{full_old_name},{old_mode},{old_type});'
+                    ' '
+                    f'new=({new_status},{full_new_name},{new_mode},{new_type})')
+                return None
+
+            new_info['type'] = candidate_type
+        return new_schema_entry
+
+    def merge_mode(self, old_schema_entry, new_schema_entry, base_path):
+        old_info = old_schema_entry['info']
```

Comment:
Let's add a docstring here, because this is a critical method, something like this:

```
This method determines if the 'mode' of a schema entry can transition from
REQUIRED -> NULLABLE. A REQUIRED mode can only have come from
an existing schema (though the --existing_schema_path flag), because REQUIRED
is created only in the flatten_schema() method. Therefore, a NULLABLE->REQUIRED
transition cannot occur.

We have the following sub cases for the REQUIRED -> NULLABLE transition:

1) If the target is filled=True, then we will retain the REQUIRED mode.
2) If the target is filled=False, then we control the outcome by overloading the
  --infer_mode flag:
  a) If --infer_mode is given, then we allow the REQUIRED -> NULLABLE transition.
  b) If --infer_mode is not given, then we log an error and ignore this field from the schema.
```

```